File: amic-platform/src/modules/ma/types/earnout.ts
export type EarnoutStatus =
  | "PENDING"
  | "MEASUREMENT_PERIOD"
  | "ACHIEVED"
  | "PARTIALLY_ACHIEVED"
  | "MISSED"
  | "DISPUTED";

export type EarnoutMetric =
  | "REVENUE"
  | "EBITDA"
  | "NET_INCOME"
  | "CUSTOMER_COUNT"
  | "CONTRACT_VALUE"
  | "WORKING_CAPITAL"
  | "OTHER";

export interface EarnoutMilestone {
  id: string;
  transaction_id: string;
  title: string;
  description: string | null;
  metric: EarnoutMetric;
  target_value: string;
  actual_value: string | null;
  currency: string;
  measurement_start: string | null;
  measurement_end: string | null;
  payment_amount: string | null;
  payment_date: string | null;
  status: EarnoutStatus;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface EarnoutCreate {
  title: string;
  description?: string | null;
  metric: EarnoutMetric;
  target_value: number;
  currency?: string;
  measurement_start?: string | null;
  measurement_end?: string | null;
  payment_amount?: number | null;
  notes?: string | null;
}

export interface EarnoutUpdate {
  title?: string;
  description?: string | null;
  metric?: EarnoutMetric;
  target_value?: number;
  actual_value?: number | null;
  currency?: string;
  measurement_start?: string | null;
  measurement_end?: string | null;
  payment_amount?: number | null;
  payment_date?: string | null;
  status?: EarnoutStatus;
  notes?: string | null;
}

export interface EarnoutSummary {
  total: number;
  total_target: string;
  total_actual: string;
  total_payment: string;
  by_status: Record<string, number>;
}

File: deal-mgmt/app/schemas/earnout.py
from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, field_serializer

from app.models.enums import EarnoutMetric, EarnoutStatus


class EarnoutOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    transaction_id: uuid.UUID
    title: str
    description: str | None = None
    metric: EarnoutMetric
    target_value: Decimal
    actual_value: Decimal | None = None
    currency: str
    measurement_start: str | None = None
    measurement_end: str | None = None
    payment_amount: Decimal | None = None
    payment_date: str | None = None
    status: EarnoutStatus
    notes: str | None = None
    created_at: datetime
    updated_at: datetime

    @field_serializer("target_value", "actual_value", "payment_amount")
    @classmethod
    def _serialize_decimal(cls, v: Decimal | None) -> str | None:
        return str(v) if v is not None else None


class EarnoutCreate(BaseModel):
    title: str = Field(..., max_length=300)
    description: str | None = None
    metric: EarnoutMetric
    target_value: Decimal = Field(..., ge=0)
    currency: str = Field("KRW", max_length=10)
    measurement_start: str | None = Field(None, max_length=10)
    measurement_end: str | None = Field(None, max_length=10)
    payment_amount: Decimal | None = Field(None, ge=0)
    notes: str | None = None


class EarnoutUpdate(BaseModel):
    title: str | None = Field(None, max_length=300)
    description: str | None = None
    metric: EarnoutMetric | None = None
    target_value: Decimal | None = Field(None, ge=0)
    actual_value: Decimal | None = Field(None, ge=0)
    currency: str | None = Field(None, max_length=10)
    measurement_start: str | None = Field(None, max_length=10)
    measurement_end: str | None = Field(None, max_length=10)
    payment_amount: Decimal | None = Field(None, ge=0)
    payment_date: str | None = Field(None, max_length=10)
    status: EarnoutStatus | None = None
    notes: str | None = None


class EarnoutSummary(BaseModel):
    total: int
    total_target: Decimal
    total_actual: Decimal
    total_payment: Decimal
    by_status: dict[str, int]

    @field_serializer("total_target", "total_actual", "total_payment")
    @classmethod
    def _serialize_decimal(cls, v: Decimal | None) -> str | None:
        return str(v) if v is not None else None
